File: swift/Sources/Rezel/Grammars/JsonGrammar.swift
import Foundation

private nonisolated(unsafe) let jsonHighlighting = styleTags([
	"String": hlString,
	"Number": hlNumber,
	"True False": (hlTags["bool"] as! Tag),
	"PropertyName": hlPropertyName,
	"Null": (hlTags["null"] as! Tag),
	", :": (hlTags["separator"] as! Tag),
	"[ ]": (hlTags["squareBracket"] as! Tag),
	"{ }": (hlTags["brace"] as! Tag),
])

public let jsonParser: LRParser = {
	let spec = JsonParserData.makeSpec(
		externals: [:],
		propSources: [jsonHighlighting]
	)
	return LRParser(spec: spec)
}()

File: dotnet/Rezel/Grammars/JsonGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class JsonGrammar
{
    private static readonly NodePropSource JsonHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["String"] = Tags.String,
        ["Number"] = Tags.Number,
        ["True False"] = Tags.Bool,
        ["PropertyName"] = Tags.PropertyName,
        ["Null"] = Tags.Null,
        [", :"] = Tags.Separator,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var spec = JsonParserData.MakeSpec(
            externals: new Dictionary<string, ITokenizer>(),
            propSources: [JsonHighlighting]
        );
        return new LRParser(spec);
    }
}
